waves-integration/src/main/scala/com/wavesplatform/dex/grpc/integration/clients/WavesBlockchainCachingClient.scala

```diff
@@ -41,7 +42,12 @@ class WavesBlockchainCachingClient(underlying: WavesBlockchainClient[Future], de
     underlying.spendableBalanceChanges
   }
 
-  override def spendableBalance(address: Address, asset: Asset): Future[Long]                           = balancesCache.get(address -> asset).map(_.toLong)
+  override def spendableBalance(address: Address, asset: Asset): Future[Long] = balancesCache.get(address -> asset).map(_.toLong)
+
+  override def allAssetsSpendableBalance(address: Address): Future[Map[Asset, Long]] = {
+    underlying.allAssetsSpendableBalance(address) unsafeTap { _.foreach(bs => balancesCache.batchPut(Map(address -> bs))) }
```

Comment:
**unsafeTap** -> **andThen**. From the documentation:

> Applies the side-effecting function to the result of this future, and returns a new future with the result of this future.

```scala
// Future.scala
  def andThen[U](pf: PartialFunction[Try[T], U])(implicit executor: ExecutionContext): Future[T] =
    transform {
      result =>
        try pf.applyOrElse[Try[T], Any](result, Predef.identity[Try[T]])
        catch { case NonFatal(t) => executor reportFailure t }

        result
    }
```

It'd better to use the standard functions instead of third-party if available.

Reply:
Done